HopsanGUI/LibraryHandler.cpp

```diff
@@ -351,6 +374,603 @@ const QVector<SharedComponentLibraryPtrT> LibraryHandler::getLibraries(const QSt
+        default:
+            spec.cqsType = "C";
+    }
+    for(int r=1; r<mpConstantsTable->rowCount(); ++r) {
+        spec.constantNames.append(mpConstantsTable->item(r,1)->text());
+        spec.constantDescriptions.append(mpConstantsTable->item(r,2)->text());
+        spec.constantUnits.append(mpConstantsTable->item(r,3)->text());
+        spec.constantInits.append(mpConstantsTable->item(r,4)->text());
+    }
+    for(int r=1; r<mpInputVariablesTable->rowCount(); ++r) {
+        spec.inputNames.append(mpInputVariablesTable->item(r,1)->text());
+        spec.inputDescriptions.append(mpInputVariablesTable->item(r,2)->text());
+        spec.inputUnits.append(mpInputVariablesTable->item(r,3)->text());
+        spec.inputInits.append(mpInputVariablesTable->item(r,4)->text());
+    }
+    for(int r=1; r<mpOutputVariablesTable->rowCount(); ++r) {
+        spec.outputNames.append(mpOutputVariablesTable->item(r,1)->text());
+        spec.outputDescriptions.append(mpOutputVariablesTable->item(r,2)->text());
+        spec.outputUnits.append(mpOutputVariablesTable->item(r,3)->text());
+        spec.outputInits.append(mpOutputVariablesTable->item(r,4)->text());
+    }
+    for(int r=1; r<mpPortsTable->rowCount(); ++r) {
+        spec.portNames.append(mpPortsTable->item(r,1)->text());
+        spec.portDescriptions.append(mpPortsTable->item(r,2)->text());
+        spec.portTypes.append(qobject_cast<QComboBox*>(mpPortsTable->cellWidget(r,3))->currentText());
+        spec.portsRequired.append(qobject_cast<QCheckBox*>(mpPortsTable->cellWidget(r,4))->isChecked());
+    }
+
+    return spec;
+}
+
+
+void NewComponentDialog::validate()
+{
+    bool error=false;
+    QBrush defaultBrush = mpGeneralTable->item(1,1)->background();
+    QBrush errorBrush(QColor(255,0,0,150));
+
+    //Clear all seletions (to make any errors visible)
+    mpGeneralTable->clearSelection();
+    mpConstantsTable->clearSelection();
+    mpInputVariablesTable->clearSelection();
+    mpOutputVariablesTable->clearSelection();
+    mpPortsTable->clearSelection();
+
+    //Reset background colors
+    mpGeneralTable->item(0,1)->setBackground(defaultBrush);
+    for(int r=1; r<mpConstantsTable->rowCount(); ++r) {
+        mpConstantsTable->item(r,1)->setBackground(defaultBrush);
+        mpConstantsTable->item(r,4)->setBackground(defaultBrush);
+    }
+    for(int r=1; r<mpInputVariablesTable->rowCount(); ++r) {
+        mpInputVariablesTable->item(r,1)->setBackground(defaultBrush);
+        mpInputVariablesTable->item(r,4)->setBackground(defaultBrush);
+    }
+    for(int r=1; r<mpOutputVariablesTable->rowCount(); ++r) {
+        mpOutputVariablesTable->item(r,1)->setBackground(defaultBrush);
+        mpOutputVariablesTable->item(r,4)->setBackground(defaultBrush);
+    }
+    for(int r=1; r<mpPortsTable->rowCount(); ++r) {
+        mpPortsTable->item(r,1)->setBackground(defaultBrush);
+    }
+
+    //Validate type name
+    ComponentSpecification spec = getSpecification();
```

Comment:
Yes, there is an `isNameValid()` function, but it does not check that the first character is a letter. We can still use it though, and check the first character separately.

Reply:
Maybe create a isTypenNameValid utility, by the way, why must the first char be a letter?
But I agre it should.

Reply:
The component name is used as class name, and according to C++ standard, class names must start with either a letter or an underscore. I decided to only allow letters in this cases, since underscore makes no sense at the beginning of a type name.

Reply:
OK makes sence